metric_learn/rca.py

```diff
@@ -61,34 +82,44 @@ def fit(self, data, chunks):
         when ``chunks[i] == -1``, point i doesn't belong to any chunklet,
         when ``chunks[i] == j``, point i belongs to chunklet j.
     """
-    data, chunks, num_chunks, d = self._process_inputs(data, chunks)
 
-    # mean center
-    data -= data.mean(axis=0)
+    data, d = self._process_data(data)
 
-    # mean center each chunklet separately
-    chunk_mask = chunks != -1
-    chunk_data = data[chunk_mask]
-    chunk_labels = chunks[chunk_mask]
-    for c in xrange(num_chunks):
-      mask = chunk_labels == c
-      chunk_data[mask] -= chunk_data[mask].mean(axis=0)
+    # PCA projection to remove noise and redundant information.
+    M_pca = None
+    if self.params['pca_comps'] is not None:
+      pca = decomposition.PCA(n_components=self.params['pca_comps'],
+                              svd_solver='full')
+      data, d = pca.fit_transform(data), self.params['pca_comps']
+      M_pca = pca.components_
 
-    # "inner" covariance of chunk deviations
+    data -= data.mean(axis=0)
+    chunk_mask, chunk_data = self._process_chunks(data, chunks)
     inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+    rank = np.linalg.matrix_rank(inner_cov)
+
+    if rank < d:
+      warnings.warn('The inner covariance matrix is not invertible, ' +
+                    'so the transformation matrix may contain Nan values. ' +
+                    'You should adjust pca_comps to remove noise and ' +
+                    'redundant information.')
```

Comment:
No need to use `+` here. The string literals will automatically concatenate without any operators between them.